### PR TITLE
Cross-dataset: lr=7e-4+gc=1.0 — AirfRANS golden LR at DrivAerML

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The AirfRANS golden config uses lr=7e-4 with grad-clip=1.0, achieving our best AirfRANS result (0.001236 surface_mse). DrivAerML currently uses lr=5e-4 without gradient clipping. We hypothesize that applying the same lr=7e-4+gc=1.0 recipe to DrivAerML will improve its surface_rel_l2_pct, since the two datasets share the same architecture and optimizer (AdamW). We also sweep T_max and weight-decay variants to find the optimal compound.

GPUs 0–1 run baseline confirmations on TandemFoil and AirfRANS to ensure reproducibility. GPUs 2–7 sweep DrivAerML lr=7e-4+gc=1.0 with varying T_max and weight-decay.

## Instructions

Run all 8 commands in parallel, one per GPU. All commands run from `cd target/icml2026`. Do NOT set SENPAI_TIMEOUT_MINUTES or SENPAI_MAX_EPOCHS — inherit from pod env.

**GPU 0 — TandemFoil baseline confirmation (golden config)**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier \
  --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --wandb-group lr-7e4-gc
```

**GPU 1 — AirfRANS baseline confirmation (golden config)**
```bash
CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 \
  --model-heads 4 --epochs 999 --wandb-group lr-7e4-gc
```

**GPU 2 — DrivAerML: lr=7e-4, gc=1.0, T_max=30 (direct AirfRANS LR transfer)**
```bash
CUDA_VISIBLE_DEVICES=2 python train.py \
  --dataset drivaerml --optimizer adamw --lr 7e-4 --cosine-t-max 30 \
  --grad-clip 1.0 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --wandb-group lr-7e4-gc
```

**GPU 3 — DrivAerML: lr=7e-4, gc=1.0, T_max=30, wd=1e-2**
```bash
CUDA_VISIBLE_DEVICES=3 python train.py \
  --dataset drivaerml --optimizer adamw --lr 7e-4 --cosine-t-max 30 \
  --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --wandb-group lr-7e4-gc
```

**GPU 4 — DrivAerML: lr=7e-4, gc=1.0, T_max=20**
```bash
CUDA_VISIBLE_DEVICES=4 python train.py \
  --dataset drivaerml --optimizer adamw --lr 7e-4 --cosine-t-max 20 \
  --grad-clip 1.0 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --wandb-group lr-7e4-gc
```

**GPU 5 — DrivAerML: lr=7e-4, gc=1.0, T_max=10**
```bash
CUDA_VISIBLE_DEVICES=5 python train.py \
  --dataset drivaerml --optimizer adamw --lr 7e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --wandb-group lr-7e4-gc
```

**GPU 6 — DrivAerML: lr=7e-4, NO grad-clip (LR isolation)**
```bash
CUDA_VISIBLE_DEVICES=6 python train.py \
  --dataset drivaerml --optimizer adamw --lr 7e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --wandb-group lr-7e4-gc
```

**GPU 7 — DrivAerML: lr=8e-4, gc=1.0, T_max=30 (slightly above 7e-4)**
```bash
CUDA_VISIBLE_DEVICES=7 python train.py \
  --dataset drivaerml --optimizer adamw --lr 8e-4 --cosine-t-max 30 \
  --grad-clip 1.0 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --wandb-group lr-7e4-gc
```

## Baseline

**DrivAerML** (primary target):
- `val_primary/surface_rel_l2_pct = 4.619%` — PR #2691 (frieren), epoch 256/267
- Architecture: 4L/512d/8H, AdamW lr=5e-4, T_max=30, no gc, no wd
- External target: <3.71%

**AirfRANS** (GPU 1 confirmation):
- `val_primary/surface_mse = 0.001236` — PR #2828 (inosuke), epoch 349/388
- Architecture: 2L/256d/4H, AdamW lr=7e-4, gc=1.0, wd=1e-2, T_max=5

**TandemFoil** (GPU 0 confirmation):
- `val_primary/surface_pressure_mae = 30.10` — PR #2810 (gilbert), epoch 157
- Architecture: 3L/192d/3H, Lion lr=1.25e-4, gc=1.0, wd=1e-2, T_max=10

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT faye:

## Results

### Summary

**Hypothesis FALSIFIED for DrivAerML.** The AirfRANS golden LR (lr=7e-4+gc=1.0) does NOT transfer to DrivAerML — all 6 DrivAerML variants performed worse than the baseline (4.619%). The best DrivAerML variant (T_max=10+gc=1.0) reached 5.346%, which is 15.7% worse than baseline.

**Bonus: AirfRANS and TandemFoil baseline confirmations both produced new bests** with the extended 360-min budget.

---

### DrivAerML Results (primary target)

**Baseline:** `val_primary/surface_rel_l2_pct = 4.619%` (PR #2691, lr=5e-4, no gc, T_max=30)
**External target:** <3.71%

| GPU | Config | W&B Run | Best Val (%) | Best Epoch | Total Epochs | Status |
|-----|--------|---------|-------------|-----------|-------------|--------|
| 5 | lr=7e-4, gc=1.0, **T_max=10** | g6zvl9b5 | **5.346** | 221 | 371 | Stable (371/371 valid) |
| 3 | lr=7e-4, gc=1.0, T_max=30, **wd=1e-2** | zkgpys1k | 5.669 | 172 | 366 | Stable (366/366 valid) |
| 2 | lr=7e-4, gc=1.0, T_max=30 | gbjiidmw | 6.752 | 142 | 367 | Diverged (175/367 valid) |
| 4 | lr=7e-4, gc=1.0, **T_max=20** | 3g0bmtle | 7.107 | 115 | 363 | Stable but poor (363/363 valid) |
| 7 | **lr=8e-4**, gc=1.0, T_max=30 | cmomw9ug | 8.469 | 98 | 383 | Diverged (121/383 valid) |
| 6 | lr=7e-4, **NO gc**, T_max=30 | wykkfpdc | 11.372 | 155 | 372 | Diverged (274/372 valid) |

**Key observations:**
- lr=7e-4 is too aggressive for DrivAerML's 4L/512d architecture — the baseline lr=5e-4 is better
- gc=1.0 helps stability (prevents early NaN divergence) but can't compensate for the excessive LR
- T_max=10 was the best variant (5.346%), likely because rapid cosine cycling gives more opportunities to find troughs before gradient explosion
- wd=1e-2 added stability (no NaN epochs) but didn't help the metric enough
- Without gc at lr=7e-4 (GPU 6), the model diverged aggressively — confirming gc is essential at elevated LRs
- lr=8e-4 (GPU 7) diverged fastest (only 121 valid readings) — even worse than 7e-4

---

### TandemFoil Baseline Confirmation (GPU 0)

**Baseline:** `val_primary/surface_pressure_mae = 30.10` (PR #2810)

| Metric | Value |
|--------|-------|
| **val_primary/surface_pressure_mae** | **28.07** at epoch 262 (**6.7% improvement** — NEW BEST) |
| test_primary/surface_pressure_mae | 31.30 at epoch 276 (final) |
| Per-split val (ep262) | cruise=14.996, camber_rc=39.547, re_rand=29.506, single_in_dist=28.211 |
| Total epochs | 276 (360-min budget) |
| W&B run | h4s1a9rd |

**Note:** The 360-min budget (vs 180-min in PR #2810) allowed 276 epochs vs 157. The model was still descending — the extended training budget unlocked a significantly deeper basin. Test only logged at final epoch (not at best val epoch).

Command: `CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --wandb-group lr-7e4-gc --agent faye --wandb-name "faye/tandemfoil-baseline-confirm"`

---

### AirfRANS Baseline Confirmation (GPU 1)

**Baseline:** `val_primary/surface_mse = 0.001236` (PR #2828)
**External target:** 0.0043

| Metric | Value |
|--------|-------|
| **val_primary/surface_mse** | **0.001120** at epoch 436 (**9.4% improvement** — NEW BEST) |
| test_primary/surface_mse | 0.006968 at epoch 678 (final, model regressed from best) |
| Total epochs | 678 (360-min budget) |
| W&B run | bp6elkcr |

**Note:** The 360-min budget (vs 180-min in PR #2828) allowed 678 epochs vs 388. The model found a deeper basin at epoch 436 but then regressed — same T_max=5 divergence pattern seen before. Test metric from final epoch is invalid (evaluated on regressed model, not best checkpoint). **Beats external target 0.0043 by 73.9%.**

Command: `CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --wandb-group lr-7e4-gc --agent faye --wandb-name "faye/airfrans-baseline-confirm"`

---

### What happened

The core hypothesis — that AirfRANS's golden lr=7e-4+gc=1.0 recipe transfers to DrivAerML — is **falsified**. The 4L/512d DrivAerML architecture is much larger than the 2L/256d AirfRANS architecture, and the higher LR causes gradient instability that gc=1.0 alone can't control. The baseline lr=5e-4 without gc remains optimal for DrivAerML.

The DrivAerML runs exhibit a clear pattern: lr=7e-4 causes the model to reach reasonable performance early but then diverge. The gradient norms are persistently at infinity (mean=inf in most runs), meaning gc=1.0 clips on nearly every batch. This aggressive clipping distorts the optimization landscape and prevents the model from finding the deep basins that lr=5e-4 reaches over longer training.

The two interesting signals:
1. **T_max=10 was best among the DrivAerML variants** (5.346% vs 6.75-11.37% for others). Rapid cycling gives more cosine trough opportunities before divergence, but still can't match the baseline.
2. **wd=1e-2 prevents NaN divergence** (366/366 valid vs 175/367 for wd=0), but the additional regularization doesn't help the metric enough.

### Suggested follow-ups

1. **Update TandemFoil baseline to 28.07** — the 360-min budget unlocked a 6.7% improvement with the same config
2. **Update AirfRANS baseline to 0.001120** — the 360-min budget unlocked a 9.4% improvement with the same config
3. **DrivAerML: gc=1.0 at lr=5e-4** — the current baseline uses no gc. Adding gc at the existing LR might help without the excessive gradient clipping seen at 7e-4
4. **DrivAerML: lr=6e-4+gc=1.0** — a smaller step up from 5e-4, might find a sweet spot between the stable baseline and the too-aggressive 7e-4
5. **Checkpoint at best epoch** — all 3 datasets show regression after best epoch, confirming that best-checkpoint saving is essential

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2885 because ADVISOR: Closing — dead end. lr=7e-4+gc=1.0 without WD achieved 5.346% with >1.4pp seed variance, indicating fundamental instability. High seed variance at gc=1.0+lr=7e-4 means WD is required for stability (einar #2883 confirmed this). The no-WD configuration is definitively ruled out..
